client/src/common/client.ts

```diff
@@ -317,6 +317,23 @@ interface _Middleware {
 	window?: WindowMiddleware;
 }
 
+// A general middleware is applied to both requests and notifications
+interface GeneralMiddleware {
+	sendRequest?<R>(
+		this: void,
+		type: string | MessageSignature,
+		next: (type: string | MessageSignature, ...params: any[]) => Promise<R>,
+		...params: any[]
```

Comment:
Actually thinking about this signature again it is not very client friendly. The reason is that the params array is hard to interpret in that form for a middleware since it can have the following combinatation:

- length 0: no param not token
- length 1: (no param and a token) or (a param no token)
- lenght 2: param and token

I have no good idea yet on how to improve this.

If all we want is to support some logging when sendRequest / sendNotification is invoked we might want to add a different API than a middleware which should allow interprestation / modification.

Reply:
@vinistock any idea regarding the above comment.

Reply:
One idea could be to have a signature like

```ts
(type: string | MessageSignature, param: any | undefined, token: CancellationToken | undefined) => Promise<R>
```

and then unfold the `params` and fold them again in the `next` handler.

Reply:
The main use case for us is for doing performance and error metrics. Something like this
```typescript
const clientOptions = {
  middleware: {
    sendRequest: () => {
      try {
        // Benchmark running the request (invoke next with params)
      } catch (error) {
        // Report error returned by the server
      }
    }
  }
}
```
For this scenario, I do think a middleware would make the most sense. It provides a lot of flexibility on what clients can do around requests and notifications.

In terms of a different API, one possibility would be to "break" the middleware into parts. This is not as flexible, and frankly feels less convenient than a middleware, but it might be easier to define an interface for. For example,
```typescript
const clientOptions = {
  beforeRequest: (type, params) => { /* start measuring performance */ },
  afterRequest: (type, params) => { /* stop measuring performance and log */ },
  onRequestError: (error, type, params) => { /* log error */ } // Could potentially be a part of `afterRequest` to avoid an extra handler
}
```

> and then unfold the params and fold them again in the next handler.

Can you explain with an example? I'm not sure I understand the approach.

Reply:
Not syntax or semantic checked but here is what I mean:

```ts
interface GeneralMiddleware {
	sendRequest?<P, R>(
		this: void,
		type: string | MessageSignature,
		param: P | undefined,
		token: CancellationToken | undefined,
		next: (type: string | MessageSignature, param: P | undefined, token?: CancellationToken) => Promise<R>,
	): Promise<R>;

	sendNotification?<P>(
		this: void,
		type: string | MessageSignature,
		params: P,
		next: (type: string | MessageSignature, params?: P) => Promise<void>
	): Promise<void>;
}
```

and in sendRequest

```ts
		const _sendRequest = this._clientOptions.middleware?.sendRequest;
		if (_sendRequest !== undefined) {
			const param: any | undefined = undefined;
			const token: CancellationToken | undefined = undefined;
			if (params.length === 1) {
				if (CancellationToken.is(params[0])) {
					token = params[0];
				} else {
					param = params[0];
				}
			} else if (params.length === 2) {
				param = params[0];
				token = params[1];
			}
			return _sendRequest(type, param, token, (type, param, token) => {
				const params: any[] = [];
				if (param !== undefined) {
					params.push(param);
				}
				if (token !== undefined) {
					params.push(token);
				}
				return connection.sendRequest<R>(type, ...params);
			});
		} else {
			return connection.sendRequest<R>(type, ...params);
		}
```

Reply:
I added your suggestion to the latest commit, so that it's easier to analyze and drop if necessary.

I wonder if there's a type signature we can come up for `params` that would improve the client interface without requiring this folding and unfolding of params. Something like
```typescript
// Params can be a CancellationToken, an `any` set of params, both of those or undefined
type ParamType = [CancellationToken] | [any] | [any, CancellationToken] |  undefined;

interface GeneralMiddleware {
	sendRequest?<R>(
		this: void,
		type: string | MessageSignature,
		next: (type: string | MessageSignature, params: ParamType) => Promise<R>,
		...params: ParamType,
	): Promise<R>;
}
```